pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/SortedInvertedIndexBasedFilterOperator.java

```diff
@@ -77,20 +77,119 @@ public BaseFilterBlock nextFilterBlock(BlockId BlockId) {
+      Arrays.sort(dictionaryIds);
+
+      IntPair lastPair = invertedIndex.getMinMaxRangeFor(dictionaryIds[0]);
+      IntRanges.clip(lastPair, startDocId, endDocId);
+
+      for (int i = 1; i < dictionaryIds.length; i++) {
+        IntPair currentPair = invertedIndex.getMinMaxRangeFor(dictionaryIds[i]);
+        IntRanges.clip(currentPair, startDocId, endDocId);
+
+        // If the previous range is degenerate, just keep the current one
+        if (IntRanges.isDegenerate(lastPair)) {
+          lastPair = currentPair;
+          continue;
+        }
+
+        // If the current range is adjacent or overlaps with the previous range, merge it into the previous range,
+        // otherwise add the previous range and keep the current one to be added
+        if (!IntRanges.rangesAreDisjoint(lastPair, currentPair)) {
+          IntRanges.mergeIntoFirst(lastPair, currentPair);
+        } else {
+          if (!IntRanges.isDegenerate(lastPair)) {
+            pairs.add(lastPair);
+          }
+          lastPair = currentPair;
+        }
+      }
+
+      // Add the last range if it's valid
+      if (!IntRanges.isDegenerate(lastPair)) {
+        pairs.add(lastPair);
+      }
+    }
+
+    if (!additiveRanges) {
+      // If the ranges are not additive ranges, our list of pairs is a list of "holes" in the [startDocId; endDocId]
+      // range. We need to take this list of pairs and invert it. To do so, there are three cases:
+      //
+      // - No holes, in which case the final range is [startDocId; endDocId]
+      // - One or more hole, in which case the final ranges are [startDocId; firstHoleStartDocId - 1] and
+      //   [lastHoleEndDocId + 1; endDocId] and ranges in between other holes
+      List<IntPair> newPairs = new ArrayList<>();
+      if (pairs.isEmpty()) {
+        newPairs.add(new IntPair(startDocId, endDocId));
       } else {
-        int newStart = Math.max(pair.getLeft(), startDocId);
-        int newEnd = Math.min(pair.getRight(), endDocId);
-        if (newStart <= newEnd) {
-          pairs.add(Pairs.intPair(newStart, newEnd));
+        // Add the first filled area (between startDocId and the first hole)
+        IntPair firstHole = pairs.get(0);
+        IntPair firstRange = new IntPair(startDocId, firstHole.getLeft() - 1);
+
+        if (!IntRanges.isDegenerate(firstRange)) {
+          newPairs.add(firstRange);
+        }
+
+        // Add the filled areas between contiguous holes
+        int pairCount = pairs.size();
+        for (int i = 1; i < pairCount; i++) {
+          IntPair previousHole = pairs.get(i - 1);
+          IntPair currentHole = pairs.get(i);
+          IntPair range = new IntPair(previousHole.getRight() + 1, currentHole.getLeft() - 1);
+          if (!IntRanges.isDegenerate(range)) {
+            newPairs.add(range);
+          }
+        }
+
+        // Add the last filled area (between the last hole and endDocId)
+        IntPair lastHole = pairs.get(pairs.size() - 1);
```

Comment:
we might have gone beyond the endDocId in the previous loop (line:177), we should check getRight is less than endDocId every time we add to newPairs